docs/codeql/codeql-language-guides/migrating-javascript-dataflow-queries.rst

```diff
@@ -0,0 +1,298 @@
+
+For example, consider this data flow configuration using a barrier guard:
+
+.. code-block:: ql
+
+    class MyConfig extends DataFlow::Configuration {
+      override predicate isBarrierGuard(DataFlow::BarrierGuardNode node) {
+        node instanceof MyBarrierGuard
+      }
+      ..
+    }
+
+    class MyBarrierGuard extends DataFlow::BarrierGuardNode {
+      MyBarrierGuard() { ... }
+
+      override predicate blocks(Expr e, boolean outcome) { ... }
+    }
+
+This can be migrated to the shared data flow library as follows:
+
+.. code-block:: ql
+
+    module MyConfig implements DataFlow::ConfigSig {
+      predicate isBarrier(DataFlow::Node node) {
+        node = DataFlow::MakeBarrierGuard<MyBarrierGuard>::getABarrierNode()
+      }
+      ..
+    }
+
+    class MyBarrierGuard extends DataFlow::Node {
+      MyBarrierGuard() { ... }
+
+      predicate blocksExpr(Expr e, boolean outcome) { ... }
+    }
+
+The changes can be summarised as:
+- The contents of ``isBarrierGuard`` have been moved to ``isBarrier``.
+- The ``node instanceof MyBarrierGuard`` check was replaced with ``node = DataFlow::MakeBarrierGuard<MyBarrierGuard>::getABarrierNode()``.
+- The ``MyBarrierGuard`` class no longer has ``DataFlow::BarrierGuardNode`` as a base class. We simply use ``DataFlow::Node`` instead.
+- The ``blocks`` predicate has been renamed to ``blocksExpr`` and no longer has the ``override`` keyword.
+
+See :ref:`using flow state <using-flow-labels-for-precise-data-flow-analysis>` for examples of how to use barrier guards with flow state.
+
+Query-specific load and store steps
+-----------------------------------
+
+The predicates ``isAdditionalLoadStep``, ``isAdditionalStoreStep``, and ``isAdditionalLoadStoreStep`` have been removed. There is no way to emulate the original behaviour.
+
+Library models can still contribute such steps, but they will be applicable to all queries. Also see the section on jump steps further down.
+
+Changes in behaviour
+--------------------
+
+When the query has been migrated to the new interface, it may seem to behave differently due to some technical differences in the internals of
+the two data flow libraries. The most significant changes are described below.
+
+Taint steps now propagate all flow states
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There's an important change from the old data flow library when using flow state and taint-tracking together.
+
+When using when using ``TaintTracking::GlobalWithState``, all flow states can propagate along taint steps.
+In the old data flow library, only the ``taint`` flow label could propagate along taint steps.
+A straight-forward translation of such a query may therefore result in new flow paths being found, which might be unexpected.
+
+To emulate the old behaviour, use ``DataFlow::GlobalWithState`` instead of ``TaintTracking::GlobalWithState``,
```

Comment:
```suggestion
To emulate the old behavior, use ``DataFlow::GlobalWithState`` instead of ``TaintTracking::GlobalWithState``,
```
british -> american